### PR TITLE
Front end development (edit button, sessions)

### PR DESCRIPTION
**Pull request by MariusKurbakovas**
Made the edit button working with backend (except for the categories part, still need to figure that out). When you edit the item it updates in UI and in the Item datbase table. Also added user session, the login is still crackable in 5 seconds so it's still shit, but it's a little better shit. Seems like this week was unfortunate for my work. On the bright side a lot of minor fixes and tweeks and Aidas added dependency injection to controllers.

### CODE REVIEW

File: GREEDY/DataManagers/ItemManager.cs [hunk 1]
```diff
@@ -41,5 +53,24 @@ public List<Item> LoadData(string Username)
...
+            using (DataBaseModel context = new DataBaseModel())
+            {
+                var itemToUpdate = context.Set<ItemDataModel>()
+                    .FirstOrDefault(x => x.ItemId == UpdatedItem.ItemId);
+                //TODO: I believe this can be written in more SOLID style
+                //Using explicit/implicit type conversion operators
+                //Didn't have the time to research this
+                itemToUpdate.Name = UpdatedItem.Name;
+                itemToUpdate.Category = UpdatedItem.Category;
+                itemToUpdate.Price = UpdatedItem.Price;
```
Comment (AidasKil) [inline GREEDY/DataManagers/ItemManager.cs:71]: Should recount the total of the receipt

**Review by AidasKil — COMMENTED** [review]
I think we should implement routing rather than have a single put and get method per controller

File: GREEDY/Controllers/CategoriesController.cs [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+            return HelperClass.JsonHttpResponse(categories);
+        }
+    }
+
+    /*
+     public async Task<HttpResponseMessage> Put()
+        {
+            HttpContent content = Request.Content;
+            string jsonContent = await content.ReadAsStringAsync();
+            //TODO: this is totaly unfinished, need to consult with the team
```
Comment (valentk777) [inline GREEDY/Controllers/CategoriesController.cs:33]: so, let's do it.

File: GREEDY/Controllers/ImagePostController.cs [hunk 3]
```diff
@@ -28,15 +35,8 @@ public HttpResponseMessage Put()
...
-            itemManager.AddItems(list, new Models.Shop() { Name = "Not supported yet", Location= "Not supported yet" }, username);
-            return new HttpResponseMessage()
-            {
-                Content = new StringContent(
-                    JsonConvert.SerializeObject(list),
-                    Encoding.UTF8,
-                    "text/html"
-                    )
-            };
+            var receiptId = _itemManager.AddItems(list, new Models.Shop() { Name = "Not supported yet", Location= "Not supported yet" }, username);
```
Comment (valentk777) [inline GREEDY/Controllers/ImagePostController.cs]: too long to read. need to start in new line

File: GREEDY/Controllers/UsersController.cs [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+        {
+            _authService = authService;
+            _sessionManager = sessionManager;
+        }
+        public async Task<HttpResponseMessage> Put()
+        {
+            HttpContent requestContent = Request.Content;
+            string jsonContent = await requestContent.ReadAsStringAsync();
+            LoginCredentials credentials = JsonConvert.DeserializeObject<LoginCredentials>(jsonContent);
+            if (credentials.Username.Length < 5 || credentials.Password.Length < 5
```
Comment (valentk777) [inline GREEDY/Controllers/UsersController.cs]: still magic numbers and not really convenient to change it fast if we would like to.

File: GREEDY/Controllers/UsersController.cs [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+            _userManager = userManager;
+        }
+        public async Task<HttpResponseMessage> Put()
+        {
+            //TODO: Change magic numbers to const
+            HttpContent requestContent = Request.Content;
+            string jsonContent = await requestContent.ReadAsStringAsync();
+            RegistrationCredentials credentials = JsonConvert.DeserializeObject<RegistrationCredentials>(jsonContent);
+            if (credentials.Username.Length < 5 || credentials.Password.Length < 5
+                || credentials.Username.Length > 256 || credentials.Password.Length > 256
```
Comment (valentk777) [inline GREEDY/Controllers/UsersController.cs]: same numbers as in Task<HttpResponseMessage> Put(). but it is okay that we have TODO task :)

File: GREEDY/DataManagers/CategoryManager.cs [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+        public List<string> GetAllDistinctCategories()
+        {
+            using (DataBaseModel context = new DataBaseModel())
+            {
+                var distinctCategories = context.Set<CategoryDataModel>()
+                    .Select(x => x.Category).Distinct();
+                return distinctCategories.ToList();
+            }
+        }
+        /*public int AddItems(IEnumerable<Item> items, Shop shop, string username)
```
Comment (valentk777) [inline GREEDY/DataManagers/CategoryManager.cs]: any comments there? why 2/3 class is commented? is it not working? or do you need time to fix it?

File: GREEDY_WEB/ClientApp/components/UserRegistration.tsx [hunk 7]
```diff
@@ -5,8 +5,8 @@ import DocumentTitle from 'react-document-title';
 import { RegistrationForm } from 'react-stormpath';
 import axios from 'axios';
 import { Link, NavLink } from 'react-router-dom';
-import Cookies from 'universal-cookie';
 import FacebookLogin from 'react-facebook-login';
```
Comment (valentk777) [inline GREEDY_WEB/ClientApp/components/UserRegistration.tsx]: fb log gin still not working and can confuse a user, but not a big problem now. :D

**Review by valentk777 — APPROVED** [review]
need to fix merge conflict and then we can merge this to master. I do not understand correctly FronEnd part, but this is not our goals in this semester. thanks for changes! user sessions is really improve our product and make users' life easier :D

**Review by valentk777 — CHANGES_REQUESTED** [review]
need to fix merge conflict, ANSWER to my question about commented code parts and then we can merge this to master. I do not understand correctly FronEnd part, but this is not our goals in this semester. thanks for changes! user sessions is really improve our product and make users' life easier :D

File: GREEDY/DataManagers/CategoryManager.cs (hunk 6 above)
Comment (MariusKurbakovas) [reply GREEDY/DataManagers/CategoryManager.cs]: Woops will fix this, this code shouldn't even be here

**Review by valentk777 — APPROVED** [review]

**Comment by valentk777** [comment]
thanks for changes. you can merge this to master :)

**Comment by valentk777** [comment]
good job. Finally we have credential parameters controller as normal class using constants

**Pull request merged**